src/software/gui/full_system/threaded_full_system_gui.cpp

```diff
@@ -91,21 +94,53 @@ void ThreadedFullSystemGUI::onValueReceived(World world)
         remaining_attempts_to_set_view_area--;
         view_area_buffer->push(world.field().fieldBoundary());
     }
+    data_per_second_buffer->push(getAverageDataPerSecond());
 }
 
 void ThreadedFullSystemGUI::onValueReceived(AIDrawFunction draw_function)
 {
     ai_draw_functions_buffer->push(draw_function);
+    data_per_second_buffer->push(getAverageDataPerSecond());
 }
 
 void ThreadedFullSystemGUI::onValueReceived(PlayInfo play_info)
 {
     play_info_buffer->push(play_info);
+    data_per_second_buffer->push(getAverageDataPerSecond());
 }
 
 void ThreadedFullSystemGUI::onValueReceived(SensorProto sensor_msg)
 {
     sensor_msg_buffer->push(sensor_msg);
+    data_per_second_buffer->push(getAverageDataPerSecond());
+}
+
+double ThreadedFullSystemGUI::getAverageDataPerSecond()
+{
+    int total = 0;
+
+    if (FirstInFirstOutThreadedObserver<World>::getDataReceivedPerSecond() > 0)
+    {
+        total++;
+    }
+    if (FirstInFirstOutThreadedObserver<AIDrawFunction>::getDataReceivedPerSecond() > 0)
+    {
+        total++;
+    }
+    if (FirstInFirstOutThreadedObserver<PlayInfo>::getDataReceivedPerSecond() > 0)
+    {
+        total++;
+    }
+    if (FirstInFirstOutThreadedObserver<SensorProto>::getDataReceivedPerSecond() > 0)
+    {
+        total++;
+    }
```

Comment:
There's duplication between these data sources, e.g. `SensorProto` causes a `World` to be produced ASAP and `AIDrawFunction` and `PlayInfo` are derivatives. So we only need to look at one data source for "vision", probably `World` is best

Also, can we have a separate data counter for `Primitive`s?

Reply:
ya, we could definitely have a `Primitive` data sending counter. I'm assuming this would require having a `ProtoMulticastListener` and networking setup?

Reply:
`Backend` is an observer of `Primitive` so it shouldn't be too hard to add

Reply:
added